vmm/src/device_manager.rs

```diff
@@ -3604,6 +3616,63 @@ impl DeviceManager {
         Ok(pci_device_bdf)
     }
 
+    fn add_pvpanic_pci_device(
+        &mut self,
+    ) -> DeviceManagerResult<Option<Arc<Mutex<devices::PvPanicDevice>>>> {
+        let id = String::from(PVPANIC_DEVICE_NAME);
+        let pci_segment_id = 0x0_u16;
+
+        info!("Creating pvpanic device {}", id);
+
+        let (pci_segment_id, pci_device_bdf, resources) =
+            self.pci_resources(&id, pci_segment_id)?;
+
+        let snapshot = snapshot_from_id(self.snapshot.as_ref(), id.as_str());
+
+        let pvpanic_device = devices::PvPanicDevice::new(id.clone(), snapshot)
+            .map_err(DeviceManagerError::PvPanicCreate)?;
+
+        let pvpanic_device = Arc::new(Mutex::new(pvpanic_device));
+
+        let new_resources = self.add_pci_device(
+            pvpanic_device.clone(),
+            pvpanic_device.clone(),
+            pci_segment_id,
+            pci_device_bdf,
+            resources,
+        )?;
+
+        let mut node = device_node!(id, pvpanic_device);
+
+        node.resources = new_resources;
+        node.pci_bdf = Some(pci_device_bdf);
+        node.pci_device_handle = None;
+
+        // Add PvPanic Device to mmio
+        self.address_manager
```

Comment:
Device BARs should already be added to the MMIO bus so this isn't necessary - if this device's BARs are not appearing then something else has gone wrong. This is done via `DeviceManager::add_pci_device` -> `PciBus::register_mapping`

Reply:
Thanks for your quick reply, Rob.